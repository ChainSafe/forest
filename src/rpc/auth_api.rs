// Copyright 2019-2023 ChainSafe Systems
// SPDX-License-Identifier: Apache-2.0, MIT

use crate::auth::*;
use crate::beacon::Beacon;
use crate::rpc_api::{auth_api::*, data_types::RPCState};
use fvm_ipld_blockstore::Blockstore;
use jsonrpc_v2::{Data, Error as JsonRpcError, Params};

/// RPC call to create a new JWT Token
pub(in crate::rpc) async fn auth_new<DB, B>(
    data: Data<RPCState<DB, B>>,
    Params(params): Params<AuthNewParams>,
) -> Result<AuthNewResult, JsonRpcError>
where
    DB: Blockstore,
    B: Beacon,
{
    let auth_params: AuthNewParams = params;
    let ks = data.keystore.read().await;
    let ki = ks.get(JWT_IDENTIFIER)?;
    let token = create_token(auth_params.perms, ki.private_key(), auth_params.token_exp)?;
    Ok(token.as_bytes().to_vec())
}

/// RPC call to verify JWT Token and return the token's permissions
pub(in crate::rpc) async fn auth_verify<DB, B>(
    data: Data<RPCState<DB, B>>,
    Params(params): Params<AuthVerifyParams>,
) -> Result<AuthVerifyResult, JsonRpcError>
where
    DB: Blockstore,
    B: Beacon,
{
    let ks = data.keystore.read().await;
    let (header_raw,) = params;
    let token = header_raw.trim_start_matches("Bearer ");
    let ki = ks.get(JWT_IDENTIFIER)?;
    let perms = verify_token(token, ki.private_key())?;
    Ok(perms)
}
