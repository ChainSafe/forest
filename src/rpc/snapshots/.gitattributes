# auto-collapse in PRs etc
/*.snap linguist-generated=true
