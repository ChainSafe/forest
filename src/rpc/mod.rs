// Copyright 2019-2023 ChainSafe Systems
// SPDX-License-Identifier: Apache-2.0, MIT

mod auth_api;
mod beacon_api;
mod chain_api;
mod common_api;
mod db_api;
mod gas_api;
mod mpool_api;
mod net_api;
mod node_api;
mod progress_api;
mod rpc_http_handler;
mod rpc_util;
mod rpc_ws_handler;
mod state_api;
mod sync_api;
mod wallet_api;

use std::{net::TcpListener, sync::Arc};

use crate::rpc_api::{
    auth_api::*, beacon_api::*, chain_api::*, common_api::*, data_types::RPCState, db_api::*,
    gas_api::*, mpool_api::*, net_api::*, node_api::NODE_STATUS, progress_api::GET_PROGRESS,
    state_api::*, sync_api::*, wallet_api::*,
};
use axum::routing::{get, post};
use fvm_ipld_blockstore::Blockstore;
use jsonrpc_v2::{Data, Error as JSONRPCError, Params, Server};
use tokio::sync::mpsc::Sender;
use tracing::info;

use crate::rpc::{
    beacon_api::beacon_get_entry,
    common_api::{shutdown, start_time, version},
    rpc_http_handler::rpc_http_handler,
    rpc_ws_handler::rpc_ws_handler,
    state_api::*,
};

pub type RpcResult<T> = Result<T, JSONRPCError>;

pub async fn start_rpc<DB>(
    state: Arc<RPCState<DB>>,
    rpc_endpoint: TcpListener,
    forest_version: &'static str,
    shutdown_send: Sender<()>,
) -> Result<(), JSONRPCError>
where
    DB: Blockstore + Clone + Send + Sync + 'static,
{
    use auth_api::*;
    use chain_api::*;
    use gas_api::*;
    use mpool_api::*;
    use sync_api::*;
    use wallet_api::*;

    let block_delay = state.state_manager.chain_config().block_delay_secs;
    let rpc_server = Arc::new(
        Server::new()
            .with_data(Data(state))
            // Auth API
            .with_method(AUTH_NEW, auth_new::<DB>)
            .with_method(AUTH_VERIFY, auth_verify::<DB>)
            // Beacon API
            .with_method(BEACON_GET_ENTRY, beacon_get_entry::<DB>)
            // Chain API
            .with_method(CHAIN_GET_MESSAGE, chain_api::chain_get_message::<DB>)
            .with_method(CHAIN_EXPORT, chain_api::chain_export::<DB>)
            .with_method(CHAIN_READ_OBJ, chain_read_obj::<DB>)
            .with_method(CHAIN_HAS_OBJ, chain_has_obj::<DB>)
            .with_method(CHAIN_GET_BLOCK_MESSAGES, chain_get_block_messages::<DB>)
            .with_method(CHAIN_GET_TIPSET_BY_HEIGHT, chain_get_tipset_by_height::<DB>)
            .with_method(CHAIN_GET_GENESIS, chain_get_genesis::<DB>)
            .with_method(CHAIN_GET_TIPSET, chain_get_tipset::<DB>)
            .with_method(CHAIN_HEAD, chain_head::<DB>)
            .with_method(CHAIN_GET_BLOCK, chain_api::chain_get_block::<DB>)
            .with_method(CHAIN_GET_NAME, chain_api::chain_get_name::<DB>)
            .with_method(CHAIN_SET_HEAD, chain_api::chain_set_head::<DB>)
            // Message Pool API
            .with_method(MPOOL_PENDING, mpool_pending::<DB>)
            .with_method(MPOOL_PUSH, mpool_push::<DB>)
            .with_method(MPOOL_PUSH_MESSAGE, mpool_push_message::<DB>)
            // Sync API
            .with_method(SYNC_CHECK_BAD, sync_check_bad::<DB>)
            .with_method(SYNC_MARK_BAD, sync_mark_bad::<DB>)
            .with_method(SYNC_STATE, sync_state::<DB>)
            // Wallet API
            .with_method(WALLET_BALANCE, wallet_balance::<DB>)
            .with_method(WALLET_DEFAULT_ADDRESS, wallet_default_address::<DB>)
            .with_method(WALLET_EXPORT, wallet_export::<DB>)
            .with_method(WALLET_HAS, wallet_has::<DB>)
            .with_method(WALLET_IMPORT, wallet_import::<DB>)
            .with_method(WALLET_LIST, wallet_list::<DB>)
            .with_method(WALLET_NEW, wallet_new::<DB>)
            .with_method(WALLET_SET_DEFAULT, wallet_set_default::<DB>)
            .with_method(WALLET_SIGN, wallet_sign::<DB>)
            .with_method(WALLET_VERIFY, wallet_verify::<DB>)
            // State API
            .with_method(STATE_CALL, state_call::<DB>)
            .with_method(STATE_REPLAY, state_replay::<DB>)
            .with_method(STATE_NETWORK_NAME, state_network_name::<DB>)
            .with_method(STATE_NETWORK_VERSION, state_get_network_version::<DB>)
            .with_method(STATE_REPLAY, state_replay::<DB>)
            .with_method(STATE_MARKET_BALANCE, state_market_balance::<DB>)
            .with_method(STATE_MARKET_DEALS, state_market_deals::<DB>)
            .with_method(STATE_GET_RECEIPT, state_get_receipt::<DB>)
            .with_method(STATE_WAIT_MSG, state_wait_msg::<DB>)
            .with_method(STATE_FETCH_ROOT, state_fetch_root::<DB>)
            // Gas API
            .with_method(GAS_ESTIMATE_FEE_CAP, gas_estimate_fee_cap::<DB>)
            .with_method(GAS_ESTIMATE_GAS_LIMIT, gas_estimate_gas_limit::<DB>)
            .with_method(GAS_ESTIMATE_GAS_PREMIUM, gas_estimate_gas_premium::<DB>)
            .with_method(GAS_ESTIMATE_MESSAGE_GAS, gas_estimate_message_gas::<DB>)
            // Common API
            .with_method(VERSION, move || version(block_delay, forest_version))
            .with_method(SHUTDOWN, move || shutdown(shutdown_send.clone()))
            .with_method(START_TIME, start_time::<DB>)
            // Net API
            .with_method(NET_ADDRS_LISTEN, net_api::net_addrs_listen::<DB>)
            .with_method(NET_PEERS, net_api::net_peers::<DB>)
            .with_method(NET_INFO, net_api::net_info::<DB>)
            .with_method(NET_CONNECT, net_api::net_connect::<DB>)
            .with_method(NET_DISCONNECT, net_api::net_disconnect::<DB>)
            // DB API
            .with_method(DB_GC, db_api::db_gc::<DB>)
            // Progress API
            .with_method(GET_PROGRESS, progress_api::get_progress)
            // Node API
            .with_method(NODE_STATUS, node_api::node_status::<DB>)
            .finish_unwrapped(),
    );

    let app = axum::Router::new()
        .route("/rpc/v0", get(rpc_ws_handler))
        .route("/rpc/v0", post(rpc_http_handler))
        .with_state(rpc_server);

    info!("Ready for RPC connections");
    let server = axum::Server::from_tcp(rpc_endpoint)?.serve(app.into_make_service());
    server.await?;

    info!("Stopped accepting RPC connections");

    Ok(())
}
