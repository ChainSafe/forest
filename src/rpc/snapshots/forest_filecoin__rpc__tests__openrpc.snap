---
source: src/rpc/mod.rs
expression: spec
---
methods:
  - name: Filecoin.ChainGetMessage
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetMessage::Result"
      schema:
        type: object
        required:
          - From
          - GasFeeCap
          - GasLimit
          - GasPremium
          - Method
          - Nonce
          - To
          - Value
          - Version
        properties:
          CID:
            $ref: "#/components/schemas/Nullable_CidLotusJsonGeneric_for_64"
          From:
            $ref: "#/components/schemas/String"
          GasFeeCap:
            $ref: "#/components/schemas/String"
          GasLimit:
            $ref: "#/components/schemas/uint64"
          GasPremium:
            $ref: "#/components/schemas/String"
          Method:
            $ref: "#/components/schemas/uint64"
          Nonce:
            $ref: "#/components/schemas/uint64"
          Params:
            $ref: "#/components/schemas/Nullable_VecU8LotusJson"
          To:
            $ref: "#/components/schemas/String"
          Value:
            $ref: "#/components/schemas/String"
          Version:
            $ref: "#/components/schemas/uint64"
      required: true
  - name: Filecoin.ChainGetParentMessages
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetParentMessages::Result"
      schema:
        type: array
        items:
          $ref: "#/components/schemas/ApiMessageLotusJson"
        nullable: true
      required: true
  - name: Filecoin.ChainGetParentReceipts
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetParentReceipts::Result"
      schema:
        type: array
        items:
          $ref: "#/components/schemas/ApiReceipt"
        nullable: true
      required: true
  - name: Filecoin.ChainGetMessagesInTipset
    params:
      - name: tsk
        schema:
          $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetMessagesInTipset::Result"
      schema:
        type: array
        items:
          $ref: "#/components/schemas/ApiMessageLotusJson"
        nullable: true
      required: true
  - name: Filecoin.ChainExport
    params:
      - name: export-params
        schema:
          type: object
          required:
            - dry_run
            - epoch
            - output_path
            - recent_roots
            - skip_checksum
            - tipset_keys
          properties:
            dry_run:
              type: boolean
            epoch:
              type: integer
              format: int64
            output_path:
              type: string
            recent_roots:
              type: integer
              format: int64
            skip_checksum:
              type: boolean
            tipset_keys:
              $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainExport::Result"
      schema:
        type: string
        nullable: true
      required: false
  - name: Filecoin.ChainReadObj
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainReadObj::Result"
      schema:
        $ref: "#/components/schemas/encoded binary"
        nullable: true
      required: true
  - name: Filecoin.ChainHasObj
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainHasObj::Result"
      schema:
        type: boolean
      required: true
  - name: Filecoin.ChainGetBlockMessages
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetBlockMessages::Result"
      schema:
        type: object
        required:
          - BlsMessages
          - Cids
          - SecpkMessages
        properties:
          BlsMessages:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::message::MessageLotusJson>"
          Cids:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>"
          SecpkMessages:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::signed_message::SignedMessageLotusJson>"
      required: true
  - name: Filecoin.ChainGetPath
    params:
      - name: from
        schema:
          $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>"
        required: true
      - name: to
        schema:
          $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetPath::Result"
      schema:
        type: array
        items:
          $ref: "#/components/schemas/PathChange_for_TipsetLotusJson"
        nullable: true
      required: true
  - name: Filecoin.ChainGetTipsetByHeight
    params:
      - name: height
        schema:
          type: integer
          format: int64
        required: true
      - name: tsk
        schema:
          type: array
          items:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_64"
          nullable: true
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetTipsetByHeight::Result"
      schema:
        type: object
        required:
          - Blocks
          - Cids
          - Height
        properties:
          Blocks:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::block_header::BlockHeaderLotusJson>"
          Cids:
            $ref: "#/components/schemas/TipsetKeyLotusJson"
          Height:
            $ref: "#/components/schemas/int64"
      required: true
  - name: Filecoin.ChainGetTipsetAfterHeight
    params:
      - name: height
        schema:
          type: integer
          format: int64
        required: true
      - name: tsk
        schema:
          type: array
          items:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_64"
          nullable: true
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetTipsetAfterHeight::Result"
      schema:
        type: object
        required:
          - Blocks
          - Cids
          - Height
        properties:
          Blocks:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::block_header::BlockHeaderLotusJson>"
          Cids:
            $ref: "#/components/schemas/TipsetKeyLotusJson"
          Height:
            $ref: "#/components/schemas/int64"
      required: true
  - name: Filecoin.ChainGetGenesis
    params: []
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetGenesis::Result"
      schema:
        $ref: "#/components/schemas/TipsetLotusJson"
        nullable: true
      required: false
  - name: Filecoin.ChainHead
    params: []
    paramStructure: by-position
    result:
      name: "Filecoin.ChainHead::Result"
      schema:
        type: object
        required:
          - Blocks
          - Cids
          - Height
        properties:
          Blocks:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::block_header::BlockHeaderLotusJson>"
          Cids:
            $ref: "#/components/schemas/TipsetKeyLotusJson"
          Height:
            $ref: "#/components/schemas/int64"
      required: true
  - name: Filecoin.ChainGetBlock
    params:
      - name: cid
        schema:
          type: object
          required:
            - /
          properties:
            /:
              $ref: "#/components/schemas/String"
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetBlock::Result"
      schema:
        type: object
        required:
          - BeaconEntries
          - ForkSignaling
          - Height
          - Messages
          - Miner
          - ParentBaseFee
          - ParentMessageReceipts
          - ParentStateRoot
          - ParentWeight
          - Parents
          - Timestamp
          - WinPoStProof
        properties:
          BLSAggregate:
            $ref: "#/components/schemas/Nullable_SignatureLotusJson"
          BeaconEntries:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::beacon_entry::BeaconEntryLotusJson>"
          BlockSig:
            $ref: "#/components/schemas/Nullable_SignatureLotusJson"
          ElectionProof:
            $ref: "#/components/schemas/Nullable_ElectionProofLotusJson"
          ForkSignaling:
            $ref: "#/components/schemas/uint64"
          Height:
            $ref: "#/components/schemas/int64"
          Messages:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
          Miner:
            $ref: "#/components/schemas/String"
          ParentBaseFee:
            $ref: "#/components/schemas/String"
          ParentMessageReceipts:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
          ParentStateRoot:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
          ParentWeight:
            $ref: "#/components/schemas/String"
          Parents:
            $ref: "#/components/schemas/TipsetKeyLotusJson"
          Ticket:
            $ref: "#/components/schemas/Nullable_TicketLotusJson"
          Timestamp:
            $ref: "#/components/schemas/uint64"
          WinPoStProof:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::po_st_proof::PoStProofLotusJson>"
      required: true
  - name: Filecoin.ChainGetTipset
    params:
      - name: tsk
        schema:
          type: array
          items:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_64"
          nullable: true
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetTipset::Result"
      schema:
        type: object
        required:
          - Blocks
          - Cids
          - Height
        properties:
          Blocks:
            $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::block_header::BlockHeaderLotusJson>"
          Cids:
            $ref: "#/components/schemas/TipsetKeyLotusJson"
          Height:
            $ref: "#/components/schemas/int64"
      required: true
  - name: Filecoin.ChainSetHead
    params:
      - name: tsk
        schema:
          type: array
          items:
            $ref: "#/components/schemas/CidLotusJsonGeneric_for_64"
          nullable: true
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainSetHead::Result"
      schema:
        type: "null"
      required: true
  - name: Filecoin.ChainGetMinBaseFee
    params:
      - name: lookback
        schema:
          type: integer
          format: uint32
          minimum: 0
        required: true
    paramStructure: by-position
    result:
      name: "Filecoin.ChainGetMinBaseFee::Result"
      schema:
        type: string
      required: true
components:
  schemas:
    ApiMessageLotusJson:
      type: object
      required:
        - Cid
        - Message
      properties:
        Cid:
          $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
        Message:
          $ref: "#/components/schemas/MessageLotusJson"
    ApiReceipt:
      type: object
      required:
        - EventsRoot
        - ExitCode
        - GasUsed
        - Return
      properties:
        EventsRoot:
          $ref: "#/components/schemas/Nullable_CidLotusJsonGeneric_for_64"
        ExitCode:
          $ref: "#/components/schemas/forest_filecoin::shim::error::ExitCode"
        GasUsed:
          type: integer
          format: uint64
          minimum: 0
        Return:
          $ref: "#/components/schemas/VecU8LotusJson2"
    BeaconEntryLotusJson:
      type: object
      required:
        - Data
        - Round
      properties:
        Data:
          $ref: "#/components/schemas/VecU8LotusJson2"
        Round:
          $ref: "#/components/schemas/uint64"
    BlockHeaderLotusJson:
      type: object
      required:
        - BeaconEntries
        - ForkSignaling
        - Height
        - Messages
        - Miner
        - ParentBaseFee
        - ParentMessageReceipts
        - ParentStateRoot
        - ParentWeight
        - Parents
        - Timestamp
        - WinPoStProof
      properties:
        BLSAggregate:
          $ref: "#/components/schemas/Nullable_SignatureLotusJson"
        BeaconEntries:
          $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::beacon_entry::BeaconEntryLotusJson>"
        BlockSig:
          $ref: "#/components/schemas/Nullable_SignatureLotusJson"
        ElectionProof:
          $ref: "#/components/schemas/Nullable_ElectionProofLotusJson"
        ForkSignaling:
          $ref: "#/components/schemas/uint64"
        Height:
          $ref: "#/components/schemas/int64"
        Messages:
          $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
        Miner:
          $ref: "#/components/schemas/String"
        ParentBaseFee:
          $ref: "#/components/schemas/String"
        ParentMessageReceipts:
          $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
        ParentStateRoot:
          $ref: "#/components/schemas/CidLotusJsonGeneric_for_642"
        ParentWeight:
          $ref: "#/components/schemas/String"
        Parents:
          $ref: "#/components/schemas/TipsetKeyLotusJson"
        Ticket:
          $ref: "#/components/schemas/Nullable_TicketLotusJson"
        Timestamp:
          $ref: "#/components/schemas/uint64"
        WinPoStProof:
          $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::po_st_proof::PoStProofLotusJson>"
    CidLotusJsonGeneric_for_64:
      type: object
      required:
        - /
      properties:
        /:
          $ref: "#/components/schemas/String"
    CidLotusJsonGeneric_for_642:
      type: object
      required:
        - /
      properties:
        /:
          $ref: "#/components/schemas/String"
    ElectionProofLotusJson:
      type: object
      required:
        - VRFProof
        - WinCount
      properties:
        VRFProof:
          $ref: "#/components/schemas/VecU8LotusJson2"
        WinCount:
          $ref: "#/components/schemas/int64"
    MessageLotusJson:
      type: object
      required:
        - From
        - GasFeeCap
        - GasLimit
        - GasPremium
        - Method
        - Nonce
        - To
        - Value
        - Version
      properties:
        CID:
          $ref: "#/components/schemas/Nullable_CidLotusJsonGeneric_for_64"
        From:
          $ref: "#/components/schemas/String"
        GasFeeCap:
          $ref: "#/components/schemas/String"
        GasLimit:
          $ref: "#/components/schemas/uint64"
        GasPremium:
          $ref: "#/components/schemas/String"
        Method:
          $ref: "#/components/schemas/uint64"
        Nonce:
          $ref: "#/components/schemas/uint64"
        Params:
          $ref: "#/components/schemas/Nullable_VecU8LotusJson"
        To:
          $ref: "#/components/schemas/String"
        Value:
          $ref: "#/components/schemas/String"
        Version:
          $ref: "#/components/schemas/uint64"
    MessageLotusJson2:
      type: object
      required:
        - From
        - GasFeeCap
        - GasLimit
        - GasPremium
        - Method
        - Nonce
        - To
        - Value
        - Version
      properties:
        CID:
          $ref: "#/components/schemas/Nullable_CidLotusJsonGeneric_for_64"
        From:
          $ref: "#/components/schemas/String"
        GasFeeCap:
          $ref: "#/components/schemas/String"
        GasLimit:
          $ref: "#/components/schemas/uint64"
        GasPremium:
          $ref: "#/components/schemas/String"
        Method:
          $ref: "#/components/schemas/uint64"
        Nonce:
          $ref: "#/components/schemas/uint64"
        Params:
          $ref: "#/components/schemas/Nullable_VecU8LotusJson"
        To:
          $ref: "#/components/schemas/String"
        Value:
          $ref: "#/components/schemas/String"
        Version:
          $ref: "#/components/schemas/uint64"
    Nullable_CidLotusJsonGeneric_for_64:
      $ref: "#/components/schemas/CidLotusJsonGeneric_for_64"
      nullable: true
    Nullable_ElectionProofLotusJson:
      $ref: "#/components/schemas/ElectionProofLotusJson"
      nullable: true
    Nullable_SignatureLotusJson:
      $ref: "#/components/schemas/SignatureLotusJson2"
      nullable: true
    Nullable_TicketLotusJson:
      $ref: "#/components/schemas/TicketLotusJson"
      nullable: true
    Nullable_VecU8LotusJson:
      $ref: "#/components/schemas/VecU8LotusJson"
      nullable: true
    PathChange_for_TipsetLotusJson:
      oneOf:
        - type: object
          required:
            - revert
          properties:
            revert:
              $ref: "#/components/schemas/TipsetLotusJson"
          additionalProperties: false
        - type: object
          required:
            - apply
          properties:
            apply:
              $ref: "#/components/schemas/TipsetLotusJson"
          additionalProperties: false
    PoStProofLotusJson:
      type: object
      required:
        - PoStProof
        - ProofBytes
      properties:
        PoStProof:
          $ref: "#/components/schemas/int64"
        ProofBytes:
          $ref: "#/components/schemas/VecU8LotusJson2"
    SignatureLotusJson:
      type: object
      required:
        - Data
        - Type
      properties:
        Data:
          $ref: "#/components/schemas/VecU8LotusJson2"
        Type:
          $ref: "#/components/schemas/SignatureTypeLotusJson"
    SignatureLotusJson2:
      type: object
      required:
        - Data
        - Type
      properties:
        Data:
          $ref: "#/components/schemas/VecU8LotusJson2"
        Type:
          $ref: "#/components/schemas/SignatureTypeLotusJson"
    SignatureType:
      description: Signature variants for Filecoin signatures.
      type: string
      enum:
        - Secp256k1
        - Bls
        - Delegated
    SignatureTypeLotusJson:
      anyOf:
        - $ref: "#/components/schemas/SignatureType"
        - $ref: "#/components/schemas/String"
    SignedMessageLotusJson:
      type: object
      required:
        - Message
        - Signature
      properties:
        CID:
          $ref: "#/components/schemas/Nullable_CidLotusJsonGeneric_for_64"
        Message:
          $ref: "#/components/schemas/MessageLotusJson"
        Signature:
          $ref: "#/components/schemas/SignatureLotusJson"
    String:
      type: string
    TicketLotusJson:
      type: object
      required:
        - VRFProof
      properties:
        VRFProof:
          $ref: "#/components/schemas/VecU8LotusJson2"
    TipsetKeyLotusJson:
      $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>"
    TipsetLotusJson:
      type: object
      required:
        - Blocks
        - Cids
        - Height
      properties:
        Blocks:
          $ref: "#/components/schemas/forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::block_header::BlockHeaderLotusJson>"
        Cids:
          $ref: "#/components/schemas/TipsetKeyLotusJson"
        Height:
          $ref: "#/components/schemas/int64"
    VecU8LotusJson:
      $ref: "#/components/schemas/encoded binary"
      nullable: true
    VecU8LotusJson2:
      $ref: "#/components/schemas/encoded binary"
      nullable: true
    encoded binary:
      type: string
    "forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::beacon_entry::BeaconEntryLotusJson>":
      type: array
      items:
        $ref: "#/components/schemas/BeaconEntryLotusJson"
      nullable: true
    "forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::block_header::BlockHeaderLotusJson>":
      type: array
      items:
        $ref: "#/components/schemas/BlockHeaderLotusJson"
      nullable: true
    "forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::cid::CidLotusJsonGeneric<64>>":
      type: array
      items:
        $ref: "#/components/schemas/CidLotusJsonGeneric_for_64"
      nullable: true
    "forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::message::MessageLotusJson>":
      type: array
      items:
        $ref: "#/components/schemas/MessageLotusJson2"
      nullable: true
    "forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::po_st_proof::PoStProofLotusJson>":
      type: array
      items:
        $ref: "#/components/schemas/PoStProofLotusJson"
      nullable: true
    "forest_filecoin::lotus_json::vec::VecLotusJson<forest_filecoin::lotus_json::signed_message::SignedMessageLotusJson>":
      type: array
      items:
        $ref: "#/components/schemas/SignedMessageLotusJson"
      nullable: true
    "forest_filecoin::shim::error::ExitCode":
      type: integer
      format: uint32
      minimum: 0
    int64:
      type: integer
      format: int64
    uint64:
      type: integer
      format: uint64
      minimum: 0
